tls/s2n_config.c

```diff
@@ -135,6 +139,55 @@ static int s2n_config_cleanup(struct s2n_config *config)
     GUARD(s2n_config_free_dhparams(config));
     GUARD(s2n_free(&config->application_protocols));
     GUARD(s2n_array_free(config->cert_and_key_pairs));
+    GUARD(s2n_map_free(config->cert_and_key_pair_map));
+
+    return 0;
+}
+
+static int s2n_config_update_cert_and_key_pair_map(struct s2n_map *cert_and_key_pair_map,
+        struct s2n_blob *name,
+        struct s2n_cert_chain_and_key *cert_key_pair)
+{
+    /* s2n_map does not allow zero-size key */
+    if (name->size == 0) {
+        return 0;
+    }
+    struct s2n_blob map_value = { 0 };
+    s2n_authentication_method auth_method = s2n_cert_chain_and_key_get_auth_method(cert_key_pair);
+    if (s2n_map_lookup(cert_and_key_pair_map, name, &map_value) == 0) {
+        struct s2n_cert_chain_and_key_map_value value = {{ 0 }};
+        value.certs[auth_method] = cert_key_pair;
+        map_value.data = (uint8_t *) &value;
+        map_value.size = sizeof(struct s2n_cert_chain_and_key_map_value);
+
+        GUARD(s2n_map_unlock(cert_and_key_pair_map));
+        GUARD(s2n_map_add(cert_and_key_pair_map, name, &map_value));
+        GUARD(s2n_map_complete(cert_and_key_pair_map));
+    } else {
+        struct s2n_cert_chain_and_key_map_value *value = (void *) map_value.data;;
+        if (value->certs[auth_method] == NULL) {
+            value->certs[auth_method] = cert_key_pair;
+        }
+    }
+
+    return 0;
+}
+
+static int s2n_config_build_cert_and_key_pair_map(struct s2n_map *cert_and_key_pair_map, struct s2n_cert_chain_and_key *cert_key_pair)
+{
+    if (s2n_array_num_elements(cert_key_pair->san_names) == 0) {
+        for (int i = 0; i < s2n_array_num_elements(cert_key_pair->cn_names); i++) {
+            struct s2n_blob *cn_name = s2n_array_get(cert_key_pair->cn_names, i);
+            GUARD(s2n_blob_char_to_lower(cn_name));
+            GUARD(s2n_config_update_cert_and_key_pair_map(cert_and_key_pair_map, cn_name, cert_key_pair));
+        }
+    } else {
+        for (int i = 0; i < s2n_array_num_elements(cert_key_pair->san_names); i++) {
+            struct s2n_blob *san_name = s2n_array_get(cert_key_pair->san_names, i);
```

Comment:
Since this is modifying the `san_names`/`cn_names` entries I think it makes more sense to do the normalization when we load the valeus in   `s2n_cert_chain_and_key_load_sans`/`s2n_cert_chain_and_key_load_cns` rather than here where we're building the map.

Reply:
make sense, done